src/Client.php

```diff
@@ -224,46 +245,46 @@ public function doRequest($request, $response = null, int $libXmlOptions = 0)
             ini_set('default_charset', 'UTF-8');
         }
 
-        $http        = $this->getHttpClient();
-        $httpRequest = $http->getRequest();
-        if ($httpRequest->getUriString() === null) {
-            $http->setUri($this->serverAddress);
-        }
-
-        $headers = $httpRequest->getHeaders();
+        $xml = $this->lastRequest->__toString();
 
-        if (! $headers->has('Content-Type')) {
-            $headers->addHeaderLine('Content-Type', 'text/xml; charset=utf-8');
+        // Build PSR-7 request
+        if ($this->requestFactory === null) {
+            throw new HttpException("No requestFactory defined");
         }
-
-        if (! $headers->has('Accept')) {
-            $headers->addHeaderLine('Accept', 'text/xml');
+        $psrRequest = $this->requestFactory
+            ->createRequest('POST', $this->serverAddress)
+            ->withHeader('Content-Type', 'text/xml; charset=utf-8')
+            ->withHeader('Accept', 'text/xml')
+            ->withHeader('User-Agent', 'Laminas_XmlRpc_Client');
+
+        if ($this->streamFactory === null) {
+            throw new HttpException("No streamFactory defined");
         }
-
-        if (! $headers->has('user-agent')) {
-            $headers->addHeaderLine('user-agent', 'Laminas_XmlRpc_Client');
+        $stream     = $this->streamFactory->createStream($xml);
+        $psrRequest = $psrRequest->withBody($stream);
+
+        try {
+            $psrResponse = $this->httpClient->sendRequest($psrRequest);
+        } catch (ClientExceptionInterface $e) {
+            // Wrap the client-specific exception in your existing HttpException
+            throw new HttpException($e->getMessage(), 0, $e);
         }
 
-        $xml = $this->lastRequest->__toString();
-        $http->setRawBody($xml);
-        $httpResponse = $http->setMethod('POST')->send();
-
-        if (! $httpResponse->isSuccess()) {
-            /**
-             * Exception thrown when an HTTP error occurs
-             */
+        $statusCode = $psrResponse->getStatusCode();
+        if ($statusCode < 200 || $statusCode >= 300) {
             throw new HttpException(
-                $httpResponse->getReasonPhrase(),
-                $httpResponse->getStatusCode()
+                $psrResponse->getReasonPhrase(),
+                $statusCode
             );
         }
 
         if ($response === null) {
             $response = new Response();
         }
```

Comment:
I don't understand why `$response` is a method argument. Why would the caller provide a response instance - only the remote endpoint can provide that!

If it can be dropped from the signature, this could be replaced with a named constructor such as `Laminas\XmlRpc\Response::fromPsr7Response($psrResponse)`.
This would be easier to test in isolation.

Also, the the response body stream is seekable, it should possibly be rewound before casting to a string.